pb/forms_test.go

```diff
@@ -0,0 +1,240 @@
+	expectedBytes := "0a050001020304"
+	expectedForm := `{
+	Data: 0x0001020304,
+	Links: nil
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Links_empty(t *testing.T) {
+	l := &PBLink{}
+	n := &PBNode{Links: []*PBLink{l}}
+	expectedBytes := "1200"
+	expectedForm := `{
+	Data: nil,
+	Links: [
+		{
+			Hash: nil,
+			Name: nil,
+			Tsize: nil
+		},
+	]
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Data_some_Links_empty(t *testing.T) {
+	l := &PBLink{}
+	n := &PBNode{Data: []byte{0, 1, 2, 3, 4}, Links: []*PBLink{l}}
+	expectedBytes := "12000a050001020304"
+	expectedForm := `{
+	Data: 0x0001020304,
+	Links: [
+		{
+			Hash: nil,
+			Name: nil,
+			Tsize: nil
+		},
+	]
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Links_zerohash(t *testing.T) {
+	l := &PBLink{Hash: make([]byte, 0)}
+	n := &PBNode{Links: []*PBLink{l}}
+	expectedBytes := "12020a00"
+	expectedForm := `{
+	Data: nil,
+	Links: [
+		{
+			Hash: 0x,
+			Name: nil,
+			Tsize: nil
+		},
+	]
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Links_zeroname(t *testing.T) {
+	name := ""
+	l := &PBLink{Name: &name}
+	n := &PBNode{Links: []*PBLink{l}}
+	expectedBytes := "12021200"
+	expectedForm := `{
+	Data: nil,
+	Links: [
+		{
+			Hash: nil,
+			Name: "",
+			Tsize: nil
+		},
+	]
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Links_zerosize(t *testing.T) {
+	var size uint64 = 0
+	l := &PBLink{Tsize: &size}
+	n := &PBNode{Links: []*PBLink{l}}
+	expectedBytes := "12021800"
+	expectedForm := `{
+	Data: nil,
+	Links: [
+		{
+			Hash: nil,
+			Name: nil,
+			Tsize: 0
+		},
+	]
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func TestForm_Links_somesize(t *testing.T) {
+	var size uint64 = 1010
+	l := &PBLink{Tsize: &size}
+	n := &PBNode{Links: []*PBLink{l}}
+	expectedBytes := "120318f207"
+	expectedForm := `{
+	Data: nil,
+	Links: [
+		{
+			Hash: nil,
+			Name: nil,
+			Tsize: 1010
+		},
+	]
+}`
+	verifyRoundTrip(t, n, expectedBytes, expectedForm)
+}
+
+func verifyRoundTrip(t *testing.T, n *PBNode, expectedBytes string, expectedForm string) {
+	actualBytes, actualForm, err := nodeRoundTripToString(n)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if actualBytes != expectedBytes {
+		fmt.Printf("Expected bytes: [%v]\nGot: [%v]\n", expectedBytes, actualBytes)
+		t.Error("Did not match")
+	}
+
+	if actualForm != expectedForm {
+		fmt.Printf("Expected form: [%v]\nGot: [%v]\n", expectedForm, actualForm)
+		t.Error("Did not match")
+	}
+}
+
+func nodeRoundTripToString(n *PBNode) (string, string, error) {
+	bytes, err := n.Marshal()
+	if err != nil {
+		return "", "", err
+	}
+	fmt.Printf("[%v]\n", hex.EncodeToString(bytes))
+	rt := new(PBNode)
+	if err := rt.Unmarshal(bytes); err != nil {
+		return "", "", err
+	}
+	return hex.EncodeToString(bytes), nodeToString(rt), nil
+}
+
+func nodeToString(n *PBNode) string {
```

Comment:
why not just use `json.MarshalIndent`? As far as I know it would do pretty much all that you need, and save you these 80 lines of rather tricky code.

I also spotted some minor issues with the code, so I think it would be easier to use JSON than to spend time in fixing it :)

Reply:
So, here's my reasoning, let's see if it makes enough sense - I wanted to be absolutely explicit about what values are set on all of these pieces, by showing `nil` where there's a nil, not just making it absent. If I switch I get this:

```
Expected form: [{
	Data: nil,
	Links: [
		{
			Hash: nil,
			Name: nil,
			Tsize: 1010
		},
	]
}]
Got: [{
	"Links": [
		{
			"Tsize": 1010
		}
	]
}]
```

I guess the two questions I have about that are - is there a 1:1 mapping of "is a `nil`" to "is absent from output" with no other ways that it might be absent? And secondly - are my "expected" chunks going to be explicit enough to express what I'm trying to demonstrate? - That 's a subjective judgement ultimately, it's about tests as documentation. But maybe it's a reasonable assumption that a user looking at this will grok that "yeah, that's missing so it must be a `nil`". I also want to pull these fixtures into JS so we're testing exactly the same thing. I could `JSON.stringify()` there and probably get the same thing but we have the added complication of `null` vs absent (`undefined`), so explicitness would be helpful.